sklearn/neighbors/binary_tree.pxi

```diff
@@ -1552,7 +1552,7 @@ cdef class BinaryTree:
         >>> import numpy as np
         >>> np.random.seed(1)
         >>> X = np.random.random((100, 3))
-        >>> tree = BinaryTree(X)           # doctest: +SKIP
+        >>> tree = {BinaryTree}(X)           # doctest: +SKIP
```

Comment:
thanks for catching this, but they're not actually being substituted in the method docstrings: https://15131-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.neighbors.BallTree.html

Reply:
I thought the line `__doc__ = CLASS_DOC.format(**DOC_DICT)` in `ball_tree.pxi` was supposed to do that, as it does for all the other occurrences of `{BinaryTree}`. Did I miss some additional step?

Reply:
It is not currently, it seems, doing it for methods of the class, only for the main class docs. Do you mind fixing that, here or in another PR?

Reply:
Not at all.

Reply:
@jnothman sorry to bother, I'm having some trouble with this. I looked around a bit and found a possible solution in the definition of a metaclass in `ball_tree.pyx` as such:
```
class SubstituteDoc(type):
    def __init__(cls, clsname, superclasses, attributedict):
        for d in dir(cls):
            if d.startswith("__"):
                continue
            dc = getattr(cls, d)
            if not callable(dc):
                continue
            getattr(cls, d).__doc__ = getattr(cls, d).__doc__.format(**DOC_DICT)
```
Which could then be applied to all subclasses:
```
# Inherit BallTree from BinaryTree
cdef class BallTree(BinaryTree,metaclass=SubstituteDoc):
    __doc__ = CLASS_DOC.format(**DOC_DICT)
    pass
```

We also would need to call this metaclass in kd_tree as well, so it should probably be defined in another file in order to be imported where needed. I haven't pushed this solution yet because I'm not sure about its cleanliness, but it's also the only one I found so far that seems to do its job.  

As this is my first contribution and I wouldn't want to mess with the quality/style of the code, could you please give me a feedback on the organisation/effectiveness of this solution?